src/frontend/src/optimizer/plan_node/logical_union.rs

```diff
@@ -0,0 +1,191 @@
+use crate::optimizer::plan_node::{BatchHashAgg, BatchUnion, LogicalAgg, PlanTreeNode};
+use crate::optimizer::property::FunctionalDependencySet;
+use crate::utils::{ColIndexMapping, Condition};
+
+/// `LogicalUnion`
+///  returns the union of the rows of its inputs
+///  if all is false means it need to eliminate duplicates.
+#[derive(Debug, Clone)]
+pub struct LogicalUnion {
+    pub base: PlanBase,
+    all: bool,
+    inputs: Vec<PlanRef>,
+}
+
+impl LogicalUnion {
+    pub fn new(all: bool, inputs: Vec<PlanRef>) -> Self {
+        let ctx = inputs[0].ctx();
+        let schema = inputs[0].schema().clone();
+        let mut pk_indices = vec![];
+        for input in &inputs {
+            for pk in input.logical_pk() {
+                if !pk_indices.contains(pk) {
+                    pk_indices.push(*pk);
+                }
+            }
+        }
+        let functional_dependency = FunctionalDependencySet::new(schema.len());
+        let base = PlanBase::new_logical(ctx, schema, pk_indices, functional_dependency);
+        LogicalUnion { base, all, inputs }
+    }
+
+    pub fn create(all: bool, inputs: Vec<PlanRef>) -> PlanRef {
+        LogicalUnion::new(all, inputs).into()
+    }
+
+    pub(super) fn fmt_with_name(&self, f: &mut fmt::Formatter, name: &str) -> fmt::Result {
+        write!(f, "{} {{ all: {} }}", name, self.all)
+    }
+
+    pub fn all(&self) -> bool {
+        self.all
+    }
+}
+
+impl PlanTreeNode for LogicalUnion {
+    fn inputs(&self) -> smallvec::SmallVec<[crate::optimizer::PlanRef; 2]> {
+        let mut vec = smallvec::SmallVec::new();
+        vec.extend(self.inputs.clone().into_iter());
+        vec
+    }
+
+    fn clone_with_inputs(&self, inputs: &[crate::optimizer::PlanRef]) -> PlanRef {
+        Self::new(self.all, inputs.to_vec()).into()
+    }
+}
+
+impl fmt::Display for LogicalUnion {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_with_name(f, "LogicalUnion")
+    }
+}
+
+impl ColPrunable for LogicalUnion {
+    fn prune_col(&self, required_cols: &[usize]) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.prune_col(required_cols))
+            .collect_vec();
+        self.clone_with_inputs(&new_inputs)
+    }
+}
+
+impl PredicatePushdown for LogicalUnion {
+    fn predicate_pushdown(&self, predicate: Condition) -> PlanRef {
+        let new_inputs = self
+            .inputs()
+            .iter()
+            .map(|input| input.predicate_pushdown(predicate.clone()))
+            .collect_vec();
+        self.clone_with_inputs(&new_inputs)
+    }
+}
+
+impl ToBatch for LogicalUnion {
+    fn to_batch(&self) -> Result<PlanRef> {
+        let new_inputs: Result<Vec<_>> =
+            self.inputs().iter().map(|input| input.to_batch()).collect();
+        let new_logical = Self::new(true, new_inputs?);
+        // convert union to union all + agg
```

Comment:
Will the `BatchUnion` executor support dedup in the future? 🤔

Reply:
I think `HashAggExecutor` is good enough, we don't need to implement the dedup logic again. BTW, union all is some kind of exchange. I think we even can use a special exchange executor to replace `UnionExecutor`.